### PR TITLE
Add webUI acceptance tests for checking set and delete profile pic (avatar)

### PR DESCRIPTION
**Pull request by dpakach**
## Description

Add webui test for checking profile picture
1. Set profile picture from uploaded files
2. Delete existing profile picture 

## Related Issue

- Related to #32850

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 1]
```diff
@@ -243,6 +243,72 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+	/**
+	 * @Then /^the preview of profile picture should (not|)\s?be shown in the webUI$/
+	 *
+	 * @param string $shouldOrNot
+	 *
+	 * @return void
+	 */
+	public function thePreviewOfProfilePictureShouldBeShownInTheWebui($shouldOrNot) {
+		$should = $shouldOrNot !== "not";
+		if ($should) {
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: This should be straightforward to check. 
```suggestion
		if ($should !== 'not') {
```

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 2]
```diff
@@ -243,6 +243,72 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+
+	/**
+	 * @Then /^the preview of profile picture should (not|)\s?be shown in the webUI$/
+	 *
+	 * @param string $shouldOrNot
+	 *
+	 * @return void
+	 */
+	public function thePreviewOfProfilePictureShouldBeShownInTheWebui($shouldOrNot) {
+		$should = $shouldOrNot !== "not";
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]:
```suggestion
```

Comment (skshetry) [reply tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: Sorry, see above comment.

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 3]
```diff
@@ -188,4 +196,69 @@ public function isGroupNameDisplayed($groupName) {
...
+
+	/**
+	 * Set profile Pic from uploaded images using the webUI
+	 *
+	 * @param string $fileName
+	 * @param Session $session
+	 *
+	 * @return void
+	 */
+	public function setProfilePic($fileName, Session $session) {
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]:
```suggestion
	public function setProfilePicture($fileName, Session $session) {
```

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 4]
```diff
@@ -188,4 +196,69 @@ public function isGroupNameDisplayed($groupName) {
...
+		$setBtn->click();
+		$this->waitForAjaxCallsToStartAndFinish($session);
+	}
+
+	/**
+	 * Check if the preview of the profile pic is shown in the webui
+	 *
+	 * @return void
+	 */
+	public function isProfilePicPreviewDisplayed() {
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]:
```suggestion
	public function isProfilePicturePreviewDisplayed() {
```

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 5]
```diff
@@ -188,4 +196,69 @@ public function isGroupNameDisplayed($groupName) {
...
+	}
+
+	/**
+	 * Delete the current profile pic
+	 *
+	 * @param Session $session
+	 *
+	 * @return void
+	 */
+	public function deleteProfilePic(Session $session) {
```
Comment (skshetry) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]:
```suggestion
	public function deleteProfilePicture(Session $session) {
```

**Review by skshetry — CHANGES_REQUESTED** [review]

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 6]
```diff
@@ -188,4 +196,69 @@ public function isGroupNameDisplayed($groupName) {
...
+	 * Set profile Pic from uploaded images using the webUI
+	 *
+	 * @param string $fileName
+	 * @param Session $session
+	 *
+	 * @return void
+	 */
+	public function setProfilePicture($fileName, Session $session) {
+		$profilePicBtn = $this->find('xpath', $this->setProfilePicFromFilesBtnXpath);
+		$this->waitTillElementIsNotNull($this->setProfilePicFromFilesBtnXpath);
```
Comment (phil-davis) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]: Should this line be before the ``find()`` ?

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 7]
```diff
@@ -188,4 +196,69 @@ public function isGroupNameDisplayed($groupName) {
...
+		$profilePicBtn->focus();
+		$profilePicBtn->click();
+		$this->waitForAjaxCallsToStartAndFinish($session);
+
+		$this->waitTillElementIsNotNull($this->setProfilePicFileListXpath);
+		$fileList = $this->find('xpath', $this->setProfilePicFileListXpath);
+
+		$this->waitTillElementIsNotNull($this->setProfilePicFileListXpath . \sprintf($this->fileListElementMatchXpath, $fileName));
```
Comment (phil-davis) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]: This line is long. Split it.

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 8]
```diff
@@ -243,6 +243,71 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
 		);
 	}
 
+	/**
+	 * @When the user sets profile picture to :filename from the uploaded files using the webUI
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: This step selects the profile picture from the user's cloud storage files - skeleton files or files they have uploaded or synced. In the test steps, we do not see any files being uploaded, but we see this step that mentions "from the uploaded files", which was a bit confusing for me.

Maybe say:
```
When the user sets the profile picture to :filename from their cloud files using the webUI
```
or?

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 9]
```diff
@@ -243,6 +243,71 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+	 * @param string $filename
+	 *
+	 * @return void
+	 */
+	public function theUserSetsProfilePictureToFromTheUploadedFiles($filename) {
+		$this->personalGeneralSettingsPage->setProfilePicture($filename, $this->getSession());
+	}
+
+	/**
+	 * @Given the user has set profile picture to :filename from the uploaded files
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: And change the words of this step to something like above:
```
Given the user has set the profile picture to :filename from their cloud files
```

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 10]
```diff
@@ -243,6 +243,71 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+	 *
+	 * @return void
+	 */
+	public function theUserHasSetProfilePictureToFromTheUploadedFiles($filename) {
+		$this->theUserSetsProfilePictureToFromTheUploadedFiles($filename);
+		$this->thePreviewOfProfilePictureShouldBeShownInTheWebui("");
+	}
+
+	/**
+	 * @Then /^the preview of profile picture should (not|)\s?be shown in the webUI$/
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: s/the preview of profile picture/the preview of the profile picture

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 11]
```diff
@@ -243,6 +243,71 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+			);
+		} else {
+			PHPUnit_Framework_Assert::assertFalse(
+				$this->personalGeneralSettingsPage->isProfilePicturePreviewDisplayed()
+			);
+		}
+	}
+
+	/**
+	 * @When the user deletes existing profile picture
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]:
```
When the user deletes the existing profile picture
```

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php [hunk 12]
```diff
@@ -243,6 +243,71 @@ public function aPasswordErrorMessageShouldBeDisplayedOnTheWebUIWithTheText(
...
+	 * @return void
+	 */
+	public function theUserDeletesExistingProfilePicture() {
+		if ($this->personalGeneralSettingsPage->isProfilePicturePreviewDisplayed()) {
+			$this->personalGeneralSettingsPage->deleteProfilePicture($this->getSession());
+		}
+	}
+
+	/**
+	 * @Given the user has deleted existing profile picture if any
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]:
```
Given the user has deleted any existing profile picture
```
reads easier IMO

File: tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature [hunk 13]
```diff
@@ -44,4 +44,14 @@ Feature: personal general settings
     Then the owncloud version should be displayed on the personal general settings page in the webUI
     And the federated cloud id for user "user1" should be displayed on the personal general settings page in the webUI
     And group "new-group" should be displayed on the personal general settings page in the webUI
-    And group "another-group" should be displayed on the personal general settings page in the webUI
\ No newline at end of file
+    And group "another-group" should be displayed on the personal general settings page in the webUI
+
+  Scenario: User sets profile picture from uploaded file
```
Comment (phil-davis) [inline tests/acceptance/features/webUIPersonalSettings/personalGeneralSettings.feature]: User sets profile picture from cloud file
?

File: tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php (hunk 8 above)
Comment (dpakach) [reply tests/acceptance/features/bootstrap/WebUIPersonalGeneralSettingsContext.php]: ok That sounds preety good to me too

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 14]
```diff
@@ -188,4 +196,75 @@ public function isGroupNameDisplayed($groupName) {
...
+		$fileList = $this->find('xpath', $this->setProfilePicFileListXpath);
+
+		$this->waitTillElementIsNotNull(
+			$this->setProfilePicFileListXpath .
+			\sprintf(
+				$this->fileListElementMatchXpath,
+				$fileName
+			)
+		);
+		$file = $fileList->find('xpath', \sprintf($this->fileListElementMatchXpath, $fileName));
```
Comment (individual-it) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]: please check also with `OwncloudPage::assertElementNotNull()`

File: tests/acceptance/features/lib/PersonalGeneralSettingsPage.php [hunk 15]
```diff
@@ -188,4 +196,75 @@ public function isGroupNameDisplayed($groupName) {
...
+	 *
+	 * @param Session $session
+	 *
+	 * @return void
+	 */
+	public function deleteProfilePicture(Session $session) {
+		$deleteBtn = $this->find('xpath', $this->profilePicDeleteBtnXpath);
+		$this->assertElementNotNull(
+			$deleteBtn,
+			__METHOD__ . "Profile Picture delete Button not found"
```
Comment (individual-it) [inline tests/acceptance/features/lib/PersonalGeneralSettingsPage.php]: space after `"`

**Review by individual-it — CHANGES_REQUESTED** [review]

**Review by skshetry — APPROVED** [review]

**Review by individual-it — APPROVED** [review]

**Comment by dpakach** [comment]
Stable10 backport on #33892

**Pull request merged**